presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java

```diff
@@ -173,22 +183,37 @@ private void onMemoryReserved()
     public boolean tryReserve(QueryId queryId, String allocationTag, long bytes)
     {
         checkArgument(bytes >= 0, "bytes is negative");
+        if (getFreeBytes() - bytes < 0) {
```

Comment:
This needs to be in the synchronized block otherwise it could allow us to allocate < 0 bytes.